### DISCUSSION TITLE
How can I use cached licenses offline?

### DISCUSSION THREAD

**Discussion by zina555** [KirillKuzyk/Google-Play-License-Check#11] [Q&A]
I don't know if this is the right place, but I have a question.

Thank you for providing us with an easy to use tool.
I would like to be able to pass for a certain period of time by means of a cached license. I would like to know how to combine the contents of BasicExample and AdvancedExample.

**Comment by KyryloKuzyk** [comment]
Hi,

Yes, this is the right place. I'm glad the plugin helps you!

> I would like to be able to pass for a certain period of time by means of a cached license.

Yes, you can use the cached license offline. If the license check was successful once in the past when the internet connection was available, you can access the cached license via the GooglePlayLicense.cachedLicense. 
There are two helper properties that you use to check if a user has a valid license:
1. GooglePlayLicense.isCachedLicenseValid. It checks if the license is LICENSED and licenseValidityTimestamp is not exceeded. licenseValidityTimestamp is typically 24 hours from the last successful connection with the licensing server.
2. GooglePlayLicense.isCachedLicenseValidOrInGracePeriod. In addition to the previous property, it also checks if a user is in the grace period. The grace period is received from the licensing server and is currently 7 days since the last successful connection to the licensing server.

If you would like to implement your own grace period, you can simply do something like this. For example, to give a user 30 days grace period:
```
static bool IsUserIn30DaysGracePeriod() {
    var cachedLicense = GooglePlayLicense.cachedLicense;
    if (cachedLicense == null) {
        return false;
    }
    // Google Play Pass recommends caching the license response for 30 days.
    return cachedLicense.status.IsLicensed() && DateTime.Now < cachedLicense.timestamp + TimeSpan.FromDays(30);
}
```
